### ISSUE TITLE
Question: Can rom be used in production?

### ISSUE THREAD

**Issue by jonahx** [rom-rb/rom#335]
I love rom's philosophy and am interested in using it on my next project.  

But I noticed on the website the section on reading aggregates said "TODO" -- and it wasn't clear if the docs were missing, or if the feature was unimplemented.  If it is just the docs that are missing, can I find examples anywhere to learn from?

Similarly, the Writing section says, "Docs are pending while we’re working on command support for rom-repository" which makes it sound like I can't write to databases using rom, but maybe I'm misunderstanding.

TLDR: Is it too early to be using this in production?

PS: What are your thoughts on Hanami's Mapper/Repository implementation?  That's another library I'm onsidering...

**Comment by solnic** [comment]
We're missing docs and a couple of convenient features (like command support for repos) but ROM is production ready in the sense that people are already using it on production.

Whether it's too early for you or not, depends solely on your needs, expectations and your willingness to get involved in the project as a contributor. ROM is not ready for the mainstream, yet, due to the lack of some convenient APIs and lots of holes in the docs. It is the priority for this year to change this situation, though...

Feature-wise it's already **very** powerful, but I would have to know what kind of project you'd like to use it for, which DB, framework etc. you're considering, in order to provide better feedback.

Regarding Hanami's model layer - it's very young and @jodosha is still exploring how its API should look like, including experimenting with ROM to see if it could be used in Hanami's model.

**Comment by jonahx** [comment]
Thanks for the reply.  It sounds like I could use it, and I'm willing to get involved and "ready for the mainstream" isn't a hard requirement.  I'll be using postgres, and my application needs are pretty standard.  My primary reason for using rom would be for a clean separation between my domain models and my persistence layer.

The main thing I'm struggling with is simply finding full worked examples within the context of a simple application (the kind of thing that will probably be in your upcoming book).  But really, I can probably get most of what I need from just a few examples (including saving a model to persistence) -- can you point me to any?  Btw, I emailed you on your website re: consulting, as I'm happy to pay for your time getting answers to some of these questions -- I know you're busy.

**Comment by solnic** [comment]
Postgres support is pretty good (we use Sequel under the hood), so that shouldn't be a problem.

I'll update rodakase-blog sample app soon and point you to some examples there. It's using my custom web-stack based on roda and dry-* gems which is a better environment for rom than rails.

**Comment by blelump** [comment]
Hi @solnic ! 

I am in the similar situation I'd say. I mean I really believe in the whole ROM concept, but on the other hand, there're doubts. I mean, the delivery time is coming quickly, whereas I am still explorling possible _edge cases_ and I'm sure the unforeseens are still ahead of me, e.g. right now I'm looking for possible authentication and authorization mechanisms under the ROM control (no success yet). 

The deal is whether to stay with warm and sweet Rails ecosystem or abandon it and start exploring, with the possibility that the delivery time would be exceeded. What would you say from your perspective?

**Comment by solnic** [comment]
@blelump I wouldn't recommend leaving Rails when you've got a project to deliver and not enough time to learn and experiment. ROM is not really ready for the so called mainstream usage. Like I said in the previous comment, if you're open to tinkering with stuff, collaborating and maybe even contributing, then it's probably the best moment, as we're still working on various core features, documentation etc. But if you've got a project to ship (esp if it's for a client) and facing some deadlines, then I'd say leave ROM for now and do it using technology that you feel comfortable with.
